checkmate/pshell.py

```diff
@@ -0,0 +1,37 @@
+import sys
+from contextlib import suppress
+
+from transaction.interfaces import NoTransaction
+
+from checkmate import models
+
+
+def setup(env):
+    sys.path = ["."] + sys.path
+
+    from tests import factories  # pylint:disable=import-outside-toplevel
+
+    sys.path = sys.path[1:]
+
+    request = env["request"]
```

Comment:
What? Where in the heck does `env` come from here?

I really don't get how this works at all

Reply:
Pyramid's `pshell` passes `env` into this function as an argument. We tell `pshell` to call this function in `development.ini`

Reply:
So we're using Pyramid's built-in interactive shell, `pshell`: https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/commandline.html#pshell-the-interactive-shell

`Makefile` runs `@tox -qe dev --run-command 'pshell conf/development.ini'`, so that's running Pyramid's built-in `pshell` script.

`pshell` will just work on its own but if you want to customize the default shell environment you can provide your own setup function to do so. So `development.ini` points `pshell` at our `setup()` function:

```ini
[pshell]
setup = checkmate.pshell.setup
```

After that it's all `pshell`: it creates its default `env` object, calls our `setup()` function passing in `env` so we can modify it, then starts a Python shell and injects that environment.

Some stuff we add on top of it:

* By running `pshell` inside tox it uses the app's version of Python (not your system version) and runs in the app's virtualenv where all the app's dependencies are available
* We install [IPython](https://ipython.org/) in that virtualenv, and I think pshell picks that up automatically and uses it instead of the default Python shell